src/ray/raylet/node_manager.cc

```diff
@@ -170,12 +170,36 @@ void HeartbeatSender::Heartbeat() {
 NodeManager::NodeManager(instrumented_io_context &io_service, const NodeID &self_node_id,
                          const NodeManagerConfig &config,
                          const ObjectManagerConfig &object_manager_config,
-                         std::shared_ptr<gcs::GcsClient> gcs_client,
-                         std::shared_ptr<ObjectDirectoryInterface> object_directory)
+                         std::shared_ptr<gcs::GcsClient> gcs_client)
     : self_node_id_(self_node_id),
       io_service_(io_service),
+      gcs_client_(gcs_client),
+      worker_pool_(io_service, self_node_id_, config.node_manager_address,
+                   config.num_workers_soft_limit,
+                   config.num_initial_python_workers_for_first_job,
+                   config.maximum_startup_concurrency, config.min_worker_port,
+                   config.max_worker_port, config.worker_ports, gcs_client_,
+                   config.worker_commands,
+                   /*starting_worker_timeout_callback=*/
+                   [this] { cluster_task_manager_->ScheduleAndDispatchTasks(); },
+                   /*get_time=*/[]() { return absl::GetCurrentTimeNanos() / 1e6; }),
+      client_call_manager_(io_service),
+      worker_rpc_pool_(client_call_manager_),
+      core_worker_subscriber_(std::make_unique<pubsub::Subscriber>(
+          self_node_id_, config.node_manager_address, config.node_manager_port,
+          RayConfig::instance().max_command_batch_size(), worker_rpc_pool_,
+          &io_service_)),
+      object_directory_(std::dynamic_pointer_cast<ObjectDirectoryInterface>(
+          std::make_shared<OwnershipBasedObjectDirectory>(
+              io_service_, gcs_client_, core_worker_subscriber_.get(),
+              [this](const ObjectID &obj_id) {
+                rpc::ObjectReference ref;
+                ref.set_object_id(obj_id.Binary());
+                MarkObjectsAsFailed(ErrorType::OBJECT_UNRECONSTRUCTABLE, {ref},
+                                    JobID::Nil());
+              }))),
```

Comment:
I know it's just copied around, but I think `std::dynamic_pointer_cast<ObjectDirectoryInterface>` is not necessary, and could you remove it?

Reply:
Done. Also changed the pointer to unique ptr